src/main/java/org/opensearch/geospatial/ip2geo/common/DatasourceManifest.java

```diff
@@ -0,0 +1,136 @@
+    private static final ParseField PROVIDER_FIELD = new ParseField("provider");
+
+    /**
+     * @param url URL of a ZIP file containing a database
+     * @return URL of a ZIP file containing a database
+     */
+    private String url;
+    /**
+     * @param dbName A database file name inside the ZIP file
+     * @return A database file name inside the ZIP file
+     */
+    private String dbName;
+    /**
+     * @param md5Hash MD5 hash value of a database file
+     * @return MD5 hash value of a database file
+     */
+    private String md5Hash;
+    /**
+     * @param validForInDays A duration in which the database file is valid to use
+     * @return A duration in which the database file is valid to use
+     */
+    private Long validForInDays;
+    /**
+     * @param updatedAt A date when the database was updated
+     * @return A date when the database was updated
+     */
+    private Long updatedAt;
+    /**
+     * @param provider A database provider name
+     * @return A database provider name
+     */
+    private String provider;
+
+    /**
+     * Ddatasource manifest parser
+     */
+    public static final ConstructingObjectParser<DatasourceManifest, Void> PARSER = new ConstructingObjectParser<>(
+        "datasource_manifest",
+        true,
+        args -> {
+            String url = (String) args[0];
+            String dbName = (String) args[1];
+            String md5hash = (String) args[2];
+            Long validForInDays = (Long) args[3];
+            Long updatedAt = (Long) args[4];
+            String provider = (String) args[5];
+            return new DatasourceManifest(url, dbName, md5hash, validForInDays, updatedAt, provider);
+        }
+    );
+    static {
+        PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), URL_FIELD);
+        PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), DB_NAME_FIELD);
+        PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), MD5_HASH_FIELD);
+        PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), VALID_FOR_IN_DAYS_FIELD);
+        PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), UPDATED_AT_FIELD);
+        PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), PROVIDER_FIELD);
+    }
+
+    /**
+     * Datasource manifest builder
+     */
+    public static class Builder {
+        private static final int MANIFEST_FILE_MAX_BYTES = 1024 * 8;
+
+        /**
+         * Build DatasourceManifest from a given url
+         *
+         * @param url url to downloads a manifest file
+         * @return DatasourceManifest representing the manifest file
+         * @throws Exception exception
+         */
+        @SuppressForbidden(reason = "Need to connect to http endpoint to read manifest file")
+        public static DatasourceManifest build(final URL url) throws Exception {
+            SpecialPermission.check();
```

Comment:
what are these special permission and where they are defined?

Reply:
Try avoiding such names.

Reply:
That is a method provided by OpenSearch core. You need to check permission before executing a code under `AccessController.doPrivileged`

Reply:
Add comment around what it is doing for better understanding.

Reply:
Add this reference here for better understanding